File: UniBlock/src/main/scala/helper/crypto/PKHelper.scala
package it.unifi.nave.uniblock
package helper.crypto

import java.nio.charset.StandardCharsets
import java.security._
import java.util.Base64
import javax.crypto.KeyAgreement

object PKHelper {
  private val PUBLIC_KEY_DH = "X25519"
  private val PUBLIC_KEY_AGREEMENT = "XDH"
  private val PUBLIC_KEY_SIGN = "Ed25519"

  def generateDhKeyPair: KeyPair = generateKeyPair(PUBLIC_KEY_DH)

  def generateSignKeyPair: KeyPair = generateKeyPair(PUBLIC_KEY_SIGN)

  private def generateKeyPair(typeKp: String) = KeyPairGenerator.getInstance(typeKp).generateKeyPair

  def encrypt(pk: PrivateKey, pbk: PublicKey, plain: Either[Array[Byte], String]): String = {
    val secret = generateSecret(pk, pbk)
    AESHelper.encrypt(secret, plain)
  }

  def decrypt(pk: PrivateKey, pbk: PublicKey, encrypted: Either[Array[Byte], String]): String = {
    val secret = generateSecret(pk, pbk)
    AESHelper.decrypt(secret, encrypted)
  }

  private def generateSecret(privateKey: PrivateKey, publicKey: PublicKey): Array[Byte] = {
    val agreement = KeyAgreement.getInstance(PUBLIC_KEY_AGREEMENT)
    agreement.init(privateKey)
    agreement.doPhase(publicKey, true)
    agreement.generateSecret
  }

  def sign(toSign: Either[Array[Byte], String], pk: PrivateKey): String = {
    val signature = Signature.getInstance(PUBLIC_KEY_SIGN)
    signature.initSign(pk)
    signature.update(toSign match {
      case Left(value) => value
      case Right(value) => value.getBytes(StandardCharsets.UTF_8)
    })
    Base64.getEncoder.encodeToString(signature.sign)
  }

  def verify(toVerify: Either[Array[Byte], String], sign: String, pbk: PublicKey): Boolean = {
    val sig = Signature.getInstance(PUBLIC_KEY_SIGN)
    sig.initVerify(pbk)
    sig.update(toVerify match {
      case Left(value) => value
      case Right(value) => value.getBytes(StandardCharsets.UTF_8)
    })
    sig.verify(Base64.getDecoder.decode(sign))
  }

}

File: program/src/main/java/it/unifi/nave/uniblock/helper/PKHelper.java
package it.unifi.nave.uniblock.helper;

import javax.crypto.KeyAgreement;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.util.Base64;

public class PKHelper {
  private static final String PUBLIC_KEY_DH = "X25519";
  private static final String PUBLIC_KEY_AGREEMENT = "XDH";
  private static final String PUBLIC_KEY_SIGN = "Ed25519";

  public static KeyPair generateDhKeyPair() {
    return generateKeyPair(PUBLIC_KEY_DH);
  }

  public static KeyPair generateSignKeyPair() {
    return generateKeyPair(PUBLIC_KEY_SIGN);
  }

  private static KeyPair generateKeyPair(String typeKp) {
    try {
      return KeyPairGenerator.getInstance(typeKp).generateKeyPair();
    } catch (NoSuchAlgorithmException e) {
      throw new RuntimeException(e);
    }
  }

  public static String encrypt(PrivateKey pk, PublicKey pbk, byte[] plain) {
    var secret = generateSecret(pk, pbk);
    return AESHelper.encrypt(secret, plain, true);
  }

  public static String decrypt(PrivateKey pk, PublicKey pbk, byte[] encrypted) {
    var secret = generateSecret(pk, pbk);
    return AESHelper.decrypt(secret, encrypted, true);
  }

  private static byte[] generateSecret(PrivateKey privateKey, PublicKey publicKey) {
    try {
      var agreement = KeyAgreement.getInstance(PUBLIC_KEY_AGREEMENT);
      agreement.init(privateKey);
      agreement.doPhase(publicKey, true);
      return agreement.generateSecret();
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  public static String sign(byte[] toSign, PrivateKey pk) {
    try {
      var signature = Signature.getInstance(PUBLIC_KEY_SIGN);
      signature.initSign(pk);
      signature.update(toSign);
      return Base64.getEncoder().encodeToString(signature.sign());
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  public static boolean verify(byte[] toVerify, String sign, PublicKey pbk) {
    try {
      var sig = Signature.getInstance(PUBLIC_KEY_SIGN);
      sig.initVerify(pbk);
      sig.update(toVerify);
      return sig.verify(Base64.getDecoder().decode(sign));
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }
}
